include/thread_pool/thread_pool.hpp

```diff
@@ -75,88 +91,129 @@ class ThreadPoolImpl {
     , m_next_worker(0)
+    , m_num_busy_waiters(0)
 {
-    for(auto& worker_ptr : m_workers)
-    {
-        worker_ptr.reset(new Worker<Task, Queue>(options.queueSize()));
-    }
+    // Instatiate all workers.
+    for (auto it = m_workers.begin(); it != m_workers.end(); ++it)
+        it->reset(new Worker<Task, Queue>(options.busyWaitOptions(), options.queueSize()));
 
-    for(size_t i = 0; i < m_workers.size(); ++i)
-    {
-        m_workers[i]->start(i, &m_workers);
-    }
-}
+    // Initialize all worker threads.
+    for (size_t i = 0; i < m_workers.size(); ++i)
+        m_workers[i]->start(i, m_workers, m_idle_workers, m_num_busy_waiters);
 
-template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>::ThreadPoolImpl(ThreadPoolImpl<Task, Queue>&& rhs) noexcept
-{
-    *this = rhs;
+    m_rouser.start(m_workers, m_idle_workers, m_num_busy_waiters);
 }
 
 template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>::~ThreadPoolImpl()
+inline GenericThreadPool<Task, Queue>::GenericThreadPool(GenericThreadPool&& rhs) noexcept
 {
-    for (auto& worker_ptr : m_workers)
-    {
-        worker_ptr->stop();
-    }
+    *this = std::move(rhs);
 }
 
 template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>&
-ThreadPoolImpl<Task, Queue>::operator=(ThreadPoolImpl<Task, Queue>&& rhs) noexcept
+inline GenericThreadPool<Task, Queue>& GenericThreadPool<Task, Queue>::operator=(GenericThreadPool&& rhs) noexcept
 {
     if (this != &rhs)
     {
+        m_idle_workers = std::move(rhs.m_idle_workers);
         m_workers = std::move(rhs.m_workers);
+        m_rouser = std::move(rhs.m_rouser);
         m_next_worker = rhs.m_next_worker.load();
+        m_num_busy_waiters = rhs.m_num_busy_waiters.load();
     }
+
     return *this;
 }
 
+template <typename Task, template<typename> class Queue>
+inline GenericThreadPool<Task, Queue>::~GenericThreadPool()
+{
+    m_rouser.stop();
+
+    for (auto& worker_ptr : m_workers)
+        worker_ptr->stop();
+}
+
 template <typename Task, template<typename> class Queue>
 template <typename Handler>
-inline bool ThreadPoolImpl<Task, Queue>::tryPost(Handler&& handler)
+inline bool GenericThreadPool<Task, Queue>::tryPost(Handler&& handler)
 {
-    return getWorker().tryPost(std::forward<Handler>(handler));
+    // This section of the code increases the probability that our thread pool
+    // is fully utilized (num active workers = argmin(num tasks, num total workers)).
+    // If there aren't busy waiters, let's see if we have any idling threads. 
+    // These incur higher overhead to wake up than the busy waiters.
+    if (m_num_busy_waiters.load(std::memory_order_acquire) == 0)
+    {
+        auto result = m_idle_workers.tryEmptyAny();
+        if (result.first)
+        {
+            auto success = m_workers[result.second]->tryPost(std::forward<Handler>(handler));
+            m_workers[result.second]->wake();
```

Comment:
if success is false, is it even worth waking the worker?

Reply:
`tryEmptyAny` removes the worker from the idle worker bag, so once we fail to post we need to ensure that the worker will be re-added to the bag. The wake function spins the worker back up so that it does this. 

Realistically this condition would be very low probability, the only time posting fails is when the queue is full, and if the worker is idle, its queue will end up empty (barring certain races).

Reply:
I think we can optimize this a little though, re-adding the worker to the queue could safely occur here as opposed to on the worker thread (via `wake()`), since the caller retains ownership of the worker's idling synchronization at this point in the execution.

Reply:
Updated to reflect https://github.com/SynaptiveMedical/thread-pool-cpp/pull/3#discussion_r174533662.